sdks/go/pkg/beam/core/runtime/exec/fn.go

```diff
@@ -117,10 +127,18 @@ func (n *invoker) Invoke(ctx context.Context, ws []typex.Window, ts typex.EventT
 	// (2) Main input from value, if any.
 	i := 0
 	if opt != nil {
-		args[in[i]] = Convert(opt.Key.Elm, fn.Param[in[i]].T)
+		if n.elmConvert == nil {
+			from := reflect.TypeOf(opt.Key.Elm)
+			n.elmConvert = ConvertFn(from, fn.Param[in[i]].T)
+		}
+		args[in[i]] = n.elmConvert(opt.Key.Elm)
```

Comment:
The change in assertion here is that all the `Elm` and `Elm2` are the same type. If that assertion isn't true, I expect the go runtime will panic. Is this a case worth testing?

Reply:
I'd rather that it panic, asserting that all values passed from the previous unit be of the same type. Most cases, the user's functions are required to be of a given type anyway, making these to be the identity function anyway. That's checked at pipeline construction time.

The cases where this isn't true are for []byte to string, since beam cheats and uses the same encoding for strings and bytes, which we don't encode separately, and for lists of "universal types" which need to be converted element by element anyway. (Universal types are the quasi-generics beam has implemented, to allow a bit of function re-use.)

If a user is returning arbitrary types as beam.T s then the runtime type checker will fail for them.